tingbot/cache.py

```diff
@@ -0,0 +1,145 @@
+#image cache
+import time
+import re
+import rfc822
+import calendar
+import requests
+import io
+import os
+from urlparse import urlparse
+
+def get_http_timestamp(dt):
+    return calendar.timegm(rfc822.parsedate(dt))
+    
+def get_server_date(response):
+    try:
+        return get_http_timestamp(response.headers['date'])
+    except (KeyError,TypeError):
+        #no date supplied, have to assume no time offset
+        return time.time()
+
+def get_last_modified(response):
+    """try and determine the last_modified time of a request"""
+    try:
+        return get_http_timestamp(response.headers['last-modified'])
+    except (KeyError,TypeError):
+        return None
+
+def get_max_age(response,last_modified):
+    """return how many seconds from original access this response is valid for"""
+    try:
+        return int(re.match(r"max-age\W*=\W*(\d+)", response.headers['cache-control']).group(1))
+    except (KeyError,ValueError,AttributeError,TypeError) as e:
+        pass
+    try:
+        return int(response.headers['max-age'])
+    except (KeyError,ValueError):
+        pass
+    try:
+        return get_http_timestamp(response.headers['expires']) - get_server_date(response)
+    except (KeyError,TypeError):
+        #really no info from server  so guess based on last-modified
+        if last_modified:
+            return min(24*60*60,(get_server_date(response)-last_modified)/10)
+        else:
+            #not even a last_modified - so conservative guess of 60s
+            return 60
+
+def get_etag(response):
+    try:
+        return response.headers['etag']
+    except KeyError:
+        return None
+
+def is_url(loc):
+    """returns true if loc is a url, and false if not"""
+    return (urlparse(loc).scheme != '')
+
+
+class ImageEntry(object):
+    #abstract base class
+    def get_image(self):
+        self.last_accessed = time.time()#local time
+        return self.image  
+        
+    def get_size(self):
+        return self.image.get_memory_usage() 
+    
+
+class WebImage(ImageEntry):
+    def __init__(self,url):
+        import graphics
+        self.url = url
+        response = requests.get(url)
+        response.raise_for_status() #raise exception if url not appropriate
+        self.set_attributes(response)
+        image_file = io.BytesIO(response.content)        
+        self.image = graphics.Image.load_file(image_file,url)
+        self.last_accessed = time.time() #local time
+        self.retrieved = time.time() #local time
+        
+    def set_attributes(self,response):    
+        self.last_modified = get_last_modified(response) #server time
```

Comment:
Actually, looking at this, we are mostly using last_modified to compare to different request responses. These may have different offsets from our local timebase (because we are comparing with when it was received - which is not necessarily when sent and is dependent on network factors). This means that out "local" last_modified may change from request to request, even though the last_modified on the server has remained the same.

For this reason, I think we should keep this as a "server time" variable (and keep it documented as such in the code)